### ISSUE TITLE
Public access dataset:DLBCL-Morphology not found in getCollection() response

### ISSUE THREAD

**Issue by gauravshetty98** [kirbyju/TCIA_Notebooks#27]
I was trying to download single images from the DLBCL-Morphology dataset but I couldn't find the collection in the getCollections() response due to which I cannot run the getSeries() function and the seriesUID. Is there some other way to access that dataset through the API or am I doing something wrong?

Dataset info: 
DOI: 10.7937/NVA3-N783
URL: https://www.cancerimagingarchive.net/collection/dlbcl-morphology

**Comment by kirbyju** [comment]
The API you're attempting to use is for our DICOM image database (NBIA).  The collection you're interested in contains pathology images that are not in DICOM format.  We use Aspera to distribute those.  https://github.com/kirbyju/TCIA_Notebooks/blob/main/TCIA_Aspera_CLI_Downloads.ipynb should help if you're trying to access these data through the command line.

**Comment by VolodymyrChapman** [comment]
Hi @gauravshetty98 ,
I can confirm that as @kirbyju  pointed out, the following functions will lead to download of the whole dataset following installation of the aspera client as pointed out in the notebook linked:

`ascli faspex5 packages receive --url='https://faspex.cancerimagingarchive.net/aspera/faspex/public/package?context=eyJyZXNvdXJjZSI6InBhY2thZ2VzIiwidHlwZSI6ImV4dGVybmFsX2Rvd25sb2FkX3BhY2thZ2UiLCJpZCI6IjYxNSIsInBhc3Njb2RlIjoiNDU1OWE3ZmJlMWJhMzkzOTQyZDJkNDFkYzRmYmFkMWYyOWY3ZjhmNyIsInBhY2thZ2VfaWQiOiI2MTUiLCJlbWFpbCI6ImhlbHBAY2FuY2VyaW1hZ2luZ2FyY2hpdmUubmV0In0=&redirected=true' `

@kirbyju Out of interest, would you know how to just receive a single file from the collection using the aspera client?
For example, just the file WSI/13952_0.svs from within the DLBCL-Morphology collection?
Thank you and best wishes,
V

**Comment by VolodymyrChapman** [comment]
Answer for my separate question about receiving single files found.
Posting in case anyone wishes to find it (the Aspera CLI documents are a nightmare to find and the best resource I managed to find was this dated guide for a severely aged version: https://download.asperasoft.com/download/docs/cli/3.7.7/user_linux/webhelp/index.html#entsrv_external/dita/ascp_examples.html). Tested and working within Google Colab (hence the prepending '!'):
`!ascli faspex5 packages receive --url='https://faspex.cancerimagingarchive.net/aspera/faspex/public/package?context=eyJyZXNvdXJjZSI6InBhY2thZ2VzIiwidHlwZSI6ImV4dGVybmFsX2Rvd25sb2FkX3BhY2thZ2UiLCJpZCI6IjYxNSIsInBhc3Njb2RlIjoiNDU1OWE3ZmJlMWJhMzkzOTQyZDJkNDFkYzRmYmFkMWYyOWY3ZjhmNyIsInBhY2thZ2VfaWQiOiI2MTUiLCJlbWFpbCI6ImhlbHBAY2FuY2VyaW1hZ2luZ2FyY2hpdmUubmV0In0=&redirected=true' /DLBCL-Morph/WSI/13952_0.svs`

Best wishes,
V

**Comment by kirbyju** [comment]
@VolodymyrChapman - Thank you sooooo much for sharing this!  I have also found their docs to be a nightmare and this is a common use case I was hoping to be able to address in the notebook eventually, but haven't had time to figure it out.  FYI, I noticed that it also works if you list more than one file or directory name with spaces in between if you want to download multiple things within a package.  E.g.

`!ascli faspex5 packages receive --url='https://faspex.cancerimagingarchive.net/aspera/faspex/public/package?context=eyJyZXNvdXJjZSI6InBhY2thZ2VzIiwidHlwZSI6ImV4dGVybmFsX2Rvd25sb2FkX3BhY2thZ2UiLCJpZCI6IjYxNSIsInBhc3Njb2RlIjoiNDU1OWE3ZmJlMWJhMzkzOTQyZDJkNDFkYzRmYmFkMWYyOWY3ZjhmNyIsInBhY2thZ2VfaWQiOiI2MTUiLCJlbWFpbCI6ImhlbHBAY2FuY2VyaW1hZ2luZ2FyY2hpdmUubmV0In0=&redirected=true' /DLBCL-Morph/WSI/13952_0.svs /DLBCL-Morph/Cells/13901/10306/1.npy /DLBCL-Morph/Cells/13901/10306/11.npy`

A couple other things I've been trying to figure out, in case anyone knows how to do this:
1. Is there some way to query the contents of a package via ascli to get a manifest so that you don't have to view the contents of the package in a browser and then make manual notes of which files you want?
2. I see there is some kind of automated checksum capability built into Aspera but I've not had any luck getting it to work.  We're currently using a separate process to generate the sums files that we put in the package.  It would be awesome if we could avoid that by using the built-in functionality.
3. Has anyone had any luck setting the options to utilize the bandwidth management features to cap the download speeds?  I tried the "-l 100m" option mentioned on that documentation page and get an error about it being an ambiguous option.

Best,
Justin

**Comment by VolodymyrChapman** [comment]
Hi @kirbyju  - glad to be of help! :-)
That's a cool feature - kudos for finding the 'chained' file requests feature.
1) I haven't investigated it in ascli but you can get a file dataframe using pathdb and then filter on say, slide properties:
```
df = pathdb.getImages(query = "DLBCL-Morphology", format = 'df')
retrieval_list = df['imageId']
file_list = [str(img)+'.svs' for img in retrieval_list]
```
I then use f-strings with os.system() to retrieve files that I want:
```
import os
for i in range(5):
    os.system(f"ascli faspex5 packages receive \
    --url='https://faspex.cancerimagingarchive.net/aspera/faspex/public/package?context=eyJyZXNvdXJjZSI6InBhY2thZ2VzIiwidHlwZSI6ImV4dGVybmFsX2Rvd25sb2FkX3BhY2thZ2UiLCJpZCI6IjYxNSIsInBhc3Njb2RlIjoiNDU1OWE3ZmJlMWJhMzkzOTQyZDJkNDFkYzRmYmFkMWYyOWY3ZjhmNyIsInBhY2thZ2VfaWQiOiI2MTUiLCJlbWFpbCI6ImhlbHBAY2FuY2VyaW1hZ2luZ2FyY2hpdmUubmV0In0=&redirected=true' \
    /DLBCL-Morph/WSI/{file_list[i]}")
```

A post request-style system would be preferred. Open to better methods if anyone has any.
My best wishes,
V

**Comment by kirbyju** [comment]
You might want to check out https://github.com/kirbyju/TCIA_Notebooks/blob/main/TCIA_PathDB_Queries.ipynb if you haven't already.  The getImages function demonstrated there will give you the URLs to download specific files from pathDB.  There are two potential limitations here though.  1) I'm not sure if the performance will be as fast on the downloads (would love feedback on that if you try it) and 2) some Aspera packages contain data that don't live anywhere in pathDB.

**Comment by VolodymyrChapman** [comment]
1) Running this small performance check:
```
import os
import time
from tcia_utils import pathdb

df = pathdb.getImages(query = "DLBCL-Morphology", format = 'df')
url = df['imageUrl'].loc[0]
filename = str(df['imageId'].loc[0]) + '.svs'
filepath = os.path.join('/DLBCL-Morph/WSI/', filename)

# wget with url
t1 = time.time()
os.system(f"wget {url}")
print('Execution time using getImages',time.time()-t1, 's')

# delete existing file -- not necessary as ascli seems to download regardless
# of existing file
if os.path.exists(filename):
  os.system(f'rm /content/{filename}')

# ascli
t2 = time.time()
os.system(f"ascli faspex5 packages receive --url='https://faspex.cancerimagingarchive.net/aspera/faspex/public/package?context=eyJyZXNvdXJjZSI6InBhY2thZ2VzIiwidHlwZSI6ImV4dGVybmFsX2Rvd25sb2FkX3BhY2thZ2UiLCJpZCI6IjYxNSIsInBhc3Njb2RlIjoiNDU1OWE3ZmJlMWJhMzkzOTQyZDJkNDFkYzRmYmFkMWYyOWY3ZjhmNyIsInBhY2thZ2VfaWQiOiI2MTUiLCJlbWFpbCI6ImhlbHBAY2FuY2VyaW1hZ2luZ2FyY2hpdmUubmV0In0=&redirected=true' {filepath}")
print('Execution time using ascli',time.time()-t2, 's')
```
In 5 repeats, wget with url takes about **27-32s** while ascli takes **14-16s**... big performance gain of ascli over large file numbers.

2) Ahhhh... right. I can see why an ascli method for retrieving a manifest would be critical now.
Just documenting here what I've tried and the errors that popped up but I'm certain that you've tried these too (all from here: https://www.rubydoc.info/gems/aspera-cli and vaguely sounding like they may retrieve a manifest or file list)

A) 
`!ascli faspex5 packages list --url="https://faspex.cancerimagingarchive.net/aspera/faspex/public/package?context=eyJyZXNvdXJjZSI6InBhY2thZ2VzIiwidHlwZSI6ImV4dGVybmFsX2Rvd25sb2FkX3BhY2thZ2UiLCJpZCI6IjYxNSIsInBhc3Njb2RlIjoiNDU1OWE3ZmJlMWJhMzkzOTQyZDJkNDFkYzRmYmFkMWYyOWY3ZjhmNyIsInBhY2thZ2VfaWQiOiI2MTUiLCJlbWFpbCI6ImhlbHBAY2FuY2VyaW1hZ2luZ2FyY2hpdmUubmV0In0=&redirected=true"` 
ERROR: Rest: Access using HTTP passcode is forbidden for this endpoint. Retry your request using an HTTP Bearer token.
status: Forbidden
code: 2009
trace: a0f414d9-b95f-4337-bf9f-44b218ec15e7
faspex.cancerimagingarchive.net 403 Forbidden

B) 
`!ascli faspex5 admin res metadata_profiles create @json:'{"name":"the profile","default":false,"title":{"max_length":200,"illegal_chars":[]},"note":{"max_length":400,"illegal_chars":[],"enabled":false},"fields":[{"ordering":0,"name":"field1","type":"text_area","require":true,"illegal_chars":[],"max_length":100},{"ordering":1,"name":"fff2","type":"option_list","require":false,"choices":["opt1","opt2"]}]}' --url="https://faspex.cancerimagingarchive.net/aspera/faspex/public/package?context=eyJyZXNvdXJjZSI6InBhY2thZ2VzIiwidHlwZSI6ImV4dGVybmFsX2Rvd25sb2FkX3BhY2thZ2UiLCJpZCI6IjYxNSIsInBhc3Njb2RlIjoiNDU1OWE3ZmJlMWJhMzkzOTQyZDJkNDFkYzRmYmFkMWYyOWY3ZjhmNyIsInBhY2thZ2VfaWQiOiI2MTUiLCJlbWFpbCI6ImhlbHBAY2FuY2VyaW1hZ2luZ2FyY2hpdmUubmV0In0=&redirected=true"`
ERROR: Rest: Access using HTTP passcode is forbidden for this endpoint. Retry your request using an HTTP Bearer token.
status: Forbidden
code: 2009
trace: b220a229-5e64-4b60-829a-4bd8687e822e
faspex.cancerimagingarchive.net 403 Forbidden

C) 
`!ascli preview show --url="https://faspex.cancerimagingarchive.net/aspera/faspex/public/package?context=eyJyZXNvdXJjZSI6InBhY2thZ2VzIiwidHlwZSI6ImV4dGVybmFsX2Rvd25sb2FkX3BhY2thZ2UiLCJpZCI6IjYxNSIsInBhc3Njb2RlIjoiNDU1OWE3ZmJlMWJhMzkzOTQyZDJkNDFkYzRmYmFkMWYyOWY3ZjhmNyIsInBhY2thZ2VfaWQiOiI2MTUiLCJlbWFpbCI6ImhlbHBAY2FuY2VyaW1hZ2luZ2FyY2hpdmUubmV0In0=&redirected=true"`
ERROR: Other(RuntimeError): missing convert binary: No such file or directory - convert
Use --log-level=debug to get more details.

**In summary:**
It appears that having a HTTP token could allow A & B to work (?), while a 'convert' bin file appears to have gone awol in time, scuppering method C.

**Comment by kirbyju** [comment]
I was able to get a hold of someone at IBM who gave me tips on listing the package files.  

Root directory of the package:
`!ascli faspex5 packages browse --url="https://faspex.cancerimagingarchive.net/aspera/faspex/public/package?context=eyJyZXNvdXJjZSI6InBhY2thZ2VzIiwidHlwZSI6ImV4dGVybmFsX2Rvd25sb2FkX3BhY2thZ2UiLCJpZCI6IjYxNSIsInBhc3Njb2RlIjoiNDU1OWE3ZmJlMWJhMzkzOTQyZDJkNDFkYzRmYmFkMWYyOWY3ZjhmNyIsInBhY2thZ2VfaWQiOiI2MTUiLCJlbWFpbCI6ImhlbHBAY2FuY2VyaW1hZ2luZ2FyY2hpdmUubmV0In0=&redirected=true"`

Listing files within subdirectories is just like with downloading multiple files.  You simply put the path after the browse command.  For example:
`!ascli faspex5 packages browse --url="https://faspex.cancerimagingarchive.net/aspera/faspex/public/package?context=eyJyZXNvdXJjZSI6InBhY2thZ2VzIiwidHlwZSI6ImV4dGVybmFsX2Rvd25sb2FkX3BhY2thZ2UiLCJpZCI6IjYxNSIsInBhc3Njb2RlIjoiNDU1OWE3ZmJlMWJhMzkzOTQyZDJkNDFkYzRmYmFkMWYyOWY3ZjhmNyIsInBhY2thZ2VfaWQiOiI2MTUiLCJlbWFpbCI6ImhlbHBAY2FuY2VyaW1hZ2luZ2FyY2hpdmUubmV0In0=&redirected=true" /DLBCL-Morph`

I will add these examples to the Aspera notebook in this repo shortly.  It would also be cool to write some kind of script that loops through the output and keeps drilling down into all subdirectories until you've got a full manifest of all the files.  That might take me a little longer to find time to do, but if anyone else has time/energy I would happily integrate it into tcia_utils so others could re-use it easily :)

**Comment by kirbyju** [comment]
Scratch the request on that last part.  I was able to put together some code to do this and for large packages like DLBCL-Morph it took 43 minutes to create the manifest!  There's no need for this at all right now since the sums file we include in each package already has a list of all the files and file paths.  My aim here was to try to get rid of the need for us (TCIA) to have to manually create and manage those.  Fortunately IBM says they're going to add a feature to easily get a full manifest in their next release of ascli, so my recommendation is to stick with using the sums files for now and just know that we'll potentially switch gears later once that new feature is added.  I'll update the notebook examples accordingly.

I also found out that Aspera has checksum functionality built into it, so if the application says something has downloaded without errors you can assume all the checksums match.

**Comment by kirbyju** [comment]
https://github.com/kirbyju/TCIA_Notebooks/blob/main/TCIA_Aspera_CLI_Downloads.ipynb is updated.  I'm going to close this out, but feel free to ping me if you run into any issues.

**Comment by ariellubonja** [comment]
> The API you're attempting to use is for our DICOM image database (NBIA). The collection you're interested in contains pathology images that are not in DICOM format. We use Aspera to distribute those. https://github.com/kirbyju/TCIA_Notebooks/blob/main/TCIA_Aspera_CLI_Downloads.ipynb should help if you're trying to access these data through the command line.


Hi @kirbyju !

Is there a way to get the metadata of all the collections at https://www.cancerimagingarchive.net/browse-collections/? I tried `nbia.getCollections()` but that returns only 125 (no account) and 143 (by logging in). However, the browse-collections website shows 214 collections in total. You mentioned Aspera, I haven't tried yet. Is all the data only stored between DICOM and Aspera?


I did a request through DataCite and that returns 290 results, which is more than the Browse Collections, which also confuses me


I would like to present this list of datasets in my lab meetings, preferably as a Spreadsheet, filterable by data type, organ, etc.



Edit: I just found `tcia_utils.wordpress`. That makes this trivial, and it indeed gives me 214 results, same as the website. Would be very helpful to make a note at [this link](https://wiki.cancerimagingarchive.net/display/Public/TCIA+Programmatic+Interface+REST+API+Guides), as the Wordpress API link redirects to the TCIA REST API, which led me to believe the two were one and the same

**Comment by kirbyju** [comment]
I made a few changes.  Seem clearer now?

**Comment by ariellubonja** [comment]
Personally, I would prefer the class name under each bullet point e.g.:

- [Collection Manager (Wordpress) REST API](https://www.cancerimagingarchive.net/collection-manager-rest-api/): TCIA has developed custom extensions to [Wordpress](https://wordpress.com/), found under `tcia_utils.wordpress` [...]
- [NBIA REST APIs](https://wiki.cancerimagingarchive.net/display/Public/National+Biomedical+Imaging+Archive+%28NBIA%29+Programmatic+Interface+REST+API+Guides): Provided as part of the NBIA software, these APIs provide access to the search and download functions used in the TCIA radiology portal, and allow access to both public and limited access DICOM collections. (`tcia_utils.nbia`)


I mean all the code is described right under, in the TCIA_UTILS paragraph, but I clicked the [Collection Manager (Wordpress) REST API](https://www.cancerimagingarchive.net/collection-manager-rest-api/) link before I got that far down

**Comment by kirbyju** [comment]
I'm a little torn about that since those guides are general API guides and are meant for anyone who wants to access them using any method or programming language (not just via python and/or tcia_utils).  I'll check with the rest of the team and see what they think.
